Get builds by status

adrianfusco opened the pull request:
- Added in Job class the --build-status argument. We can use it to filter the jobs builds by status.
- Added in the AttributeDictValue class the __delitem__ method. In this case we can delete the elements by key in the same dictionary without creating a new one.
- Added the method get_builds_by_status in the Elasticsearch class. We will use it to filter the jobs builds by status.
- Added method to test get_builds_by_status.

cescgina commented on cibyl/models/ci/job.py:
```diff
@@ -48,7 +48,10 @@ class Job(Model):
                                    func='get_last_build', nargs=0,
-                                   description="Last build for job")]
+                                   description="Last build for job"),
+                          Argument(name='--build-status', arg_type=str,
```
There is an identical Argument already in Build, is there any reason to add it here?

bregman-arie replied:
+1 to what Joan wrote. It's a build attribute, so it fits better in the Build model.

adrianfusco replied:
It makes sense, thank you Joan and Arie. In this case I should move the logic to the `get_builds` method. I'll do the change.

adrianfusco replied:
Changed :+1:

cescgina commented on cibyl/sources/elasticsearch/api.py:
```diff
@@ -47,7 +47,7 @@ def __init__(self: object, driver: str = 'elasticsearch',
             self.es_client = ElasticSearchClient(host, port).connect()
 
-    def get_jobs(self: object, **kwargs) -> list:
+    def get_jobs(self: object, **kwargs: str) -> list:
```
The type annotation here should be dict, not str, shouldn't it?

adrianfusco replied:
In this case we're receiving dictionaries with string keys, shouldn't it be **string** in this case? Reference: https://peps.python.org/pep-0484/#arbitrary-argument-lists-and-default-argument-values

Maybe I'm confusing with this. Let me know your opinion, please.

cescgina replied:
I didn't know that, so you're right it should not be dict. However, according that PEP, the type annotated is the value, not the keys. So it should be Argument, not str.

adrianfusco replied:
It makes sense. I’ll do the change later. Thanks!

adrianfusco replied:
Changed.

cescgina commented on cibyl/sources/elasticsearch/api.py:
```diff
@@ -144,6 +144,27 @@ def get_last_build(self, **kwargs):
 
         return AttributeDictValue("jobs", attr_type=Job, value=job_object)
 
+    def get_builds_by_status(self: object, **kwargs: str):
```
I would use the general get_builds method for this case too. The user might want to filter by build id and status at the same time, so calling two different functions might give the wrong results if not all filters are applied at once.

adrianfusco replied:
Mmm. In this case the method I’ve done uses `get_builds`, it loops over  the builds and delete of the dictionary all the builds that doesn’t match the status. Idk if it’s better to add more logic into the `get_builds` method

adrianfusco replied:
I'll do the change, I've answered this one in: https://github.com/rhos-infra/cibyl/pull/97#discussion_r833809978 and continue there. Thx Joan.

cescgina reviewed (commented):
Left a few comments

adrianfusco commented:
- Change type hint from str to Argument in kwargs.
- Delete method get_builds_by_status. The logic is now in get_builds.
- Deleted Argument build-status from Job class. It is already in Build.
- Changed tests to apply this new case.

bregman-arie reviewed (approved):
Looks good! :)

cescgina reviewed (approved):
Looks great, good job!
